### PR TITLE
#7723 - Adding Support For '0x' Notation Color Format

### PR DESCRIPTION
**Pull request by nyteksf**
Hello. Could somebody please point me in the right direction from here where I am stuck? I have implemented the RegExp used to find "0x..." format colors which are entered as input. I have also added the associated button, and plugged in the related code that it might be used (for the most part). However, I'm now having trouble linking everything up in a non-hackish way (or at all) in large part because tinycolor doesn't recognize 0x notation. As a result I cannot get the color swatches to work, or 0x notation to be easily recognized by the given methods of tinycolor like .getFormat() which is for example used with a switch case in Brackets for control flow.  Thanks in advance to anyone who can lend me their support! :)

@zaggino

### CODE REVIEW

**Comment by petetnt** [comment]
Hi @nyteksf,

if the issue is that tinycolor doesn't support `0x` format, I would suggest that you _always_ convert any potential `0x` values to RGB first. Basically kinda like you have at your `checkIf0xNotation`, but much more simpler. Something like:

``` js
function asRGB(color) {
  if (/0x/.test(color)) {
    return color.replace("0x", "#");
  }

  return color;
}

/** somewhere down the line where tinycolor is called **/
tinycolor(asRGB(this._getColor())

// when and only when you need the actual color value (for example presentation) convert it to 0x format: 

if (this._format === "0x") {
  return color.toHexString().replace("#", "0x");
}

return color.toHexString();
```

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: should really be `/^0x/.test(color)`

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: instead of `thatColor` use something more appropriate, like `hexColor`

**Review by zaggino — CHANGES_REQUESTED** [review]
doesn't look that bad, as @petetnt said, define helper functions like `isColor0xNotation`, `color0xToHex`, `colorHexTo0x` and use them in your code

**Comment by ficristo** [comment]
@nyteksf if TinyColor doesn't support the 0x notation, I think you should file an issue to https://github.com/bgrins/TinyColor to ask support for it.
(From what I saw newer version of the library still don't support the 0x notation, but you could double check)
ATM that repo doesn't seem very active so maybe nothing will come out anytime soon, but if they aren't aware of the issue it will never be fixed.

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: do not create a variable if it's used only once, you can do `return color0xToHex(color,toStr)` directly

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: when `if` branch ends with a return statement, else branch is not required and introduces unnecessary nesting, write
```
if (condition) {
  return something;
}
return somethingelse;
```
instead of 
```
if (condition) {
  return something;
} else {
  return somethingelse;
}
```

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: what does this do?
I'd expect `toString();` result to be assigned to something?
or did you mean to do `return hexColor.toString();` ?

Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: `return hexColor.toString()` is correct.

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Nit: space between `} else {`

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: this `newColor.replace("0x","#") || newColor` should be wrapped in a function with a proper name like `normalizeColorString` or something which hints, why we're doing this

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Nit: spaces: ` if ((/^0x/).test(color)) {`

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: again, wrap this in some sort of function

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Not a huge fan of the second parameter, maybe rename it to something like `convertToString` and make it a boolean 🤔

File: src/extensions/default/QuickView/main.js
Comment (zaggino) [inline src/extensions/default/QuickView/main.js]: this ending bracket `}` seems to be wrongly indented

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Nits: Can be written as `if (!toStr)`

File: src/extensions/default/QuickView/main.js
Comment (petetnt) [inline src/extensions/default/QuickView/main.js]: Nit: spaces `if ((/^0x/).test(previewCSS)) {`, misaligned bracket on line 407

**Comment by zaggino** [comment]
@nyteksf added a few comments, please have a look

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: See the comment about the second paremeter above

**Review by petetnt — CHANGES_REQUESTED** [review]
Couple of simple comments, looking good so far! Keep up the good work!

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: You meant `if (toStr)` :)

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (nyteksf) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Aha! That's the idea behind what I was going for. Thanks. :)

**Comment by nyteksf** [comment]
Okay. Firstly, I just want to take a moment to thank you guys, once again, for helping me--each one of these feedback sessions has been worth at least a million articles or book pages by contrast.

That said, in response to my most recent changes, I think I have satisfied your associated change requests. Or no? Please let me know if otherwise, and again, I will hop to that so I can get onto the next one. Good times. :)

File: src/extensions/default/QuickView/main.js
Comment (zaggino) [inline src/extensions/default/QuickView/main.js]: can this be a function like:
```
function ensureHexFormat(str) {
  return (/^0x/).test(str) ? str.replace("0x","#") : str;
}
```
adding constructs like these without explanation makes the code less readable to me

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: you can re-use the `ensureHexFormat` function I mentioned in the comment before here?

**Comment by zaggino** [comment]
Just two of my comments, lets wait for @petetnt to re-check this too.

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: One more nit: `if (convertToString) {` (we really should enable http://eslint.org/docs/rules/space-after-keywords and similar in our config, our spaces are wildly inconsistent across the codebase 🤔 )

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Nit: space after comma `color, convertToStr`

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (petetnt) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: Nit: space after comma `color, convertToStr`

**Review by petetnt — COMMENTED** [review]
Couple of small nits from me + @zaggino's comments, otherwise LGTM. We need to add some tests for this in another commit (and check that the relevant tests still pass)

File: src/extensions/default/InlineColorEditor/ColorEditor.js
Comment (zaggino) [inline src/extensions/default/InlineColorEditor/ColorEditor.js]: @petetnt https://github.com/adobe/brackets/pull/13177

**Comment by nyteksf** [comment]
Okay. The latest change requests have been fulfilled, or so I believe. As always--please don't hesitate to let me know if there're any more in mind. Otherwise, thanks again. And see you again soon. :)

@zaggino @petetnt

File: src/extensions/default/QuickView/main.js
Comment (zaggino) [inline src/extensions/default/QuickView/main.js]: @nyteksf 
wrong indentation, do you have `brackets-eslint` extension installed? it should have given you a warning here

**Comment by zaggino** [comment]
Looks good now except one place where the code is incorrectly indented, it'd be a good thing to install `brackets-eslint` extension and fix the warnings in the code you've modified (you don't need to fix warnings in code you didn't touch of course, we know there're some, but we're trying to get the codebase to be more consistent)

File: src/extensions/default/QuickView/main.js
Comment (nyteksf) [inline src/extensions/default/QuickView/main.js]: Whoops. That said, it passed my tests. And I had thought I had brackets-eslint installed, but I'm not sure anymore. Is that not the same as what would be found when running the command `grunt eslint`? I ask as it seems npm just downloaded something. If not, then it should be taken care of from here on out, at least.

**Review by petetnt — APPROVED** [review]
I added some missing JSDoc strings, otherwise LGTM if tests pass.

**Comment by nyteksf** [comment]
:)

@petetnt

**Review by zaggino — APPROVED** [review]

**Comment by zaggino** [comment]
Merged, thanks @nyteksf

**Comment by petetnt** [comment]
Great job @nyteksf! Thanks for this contribution!

**Pull request merged**
